src/sentry/services/hybrid_cloud/rpc.py

```diff
@@ -0,0 +1,270 @@
+from __future__ import annotations
+
+import inspect
+import logging
+from abc import abstractmethod
+from typing import Any, Callable, Dict, Iterator, Mapping, Tuple, Type, TypeVar, cast
+
+import pydantic
+
+from sentry.services.hybrid_cloud import ArgumentDict, DelegatedBySiloMode, InterfaceWithLifecycle
+from sentry.silo import SiloMode
+from sentry.types.region import Region
+
+logger = logging.getLogger(__name__)
+
+_T = TypeVar("_T")
+
+_IS_RPC_METHOD_ATTR = "__is_rpc_method"
+
+
+class RpcServiceSetupException(Exception):
+    """Indicates an error in declaring the properties of RPC services."""
+
+
+class RpcMethodSignature:
+    """Represent the set of parameters expected for one RPC method.
+
+    This class is responsible for serializing and deserializing arguments. If the
+    base service runs in the region silo, this class is also responsible for
+    resolving the arguments to the correct region for a remote call.
+    """
+
+    def __init__(self, base_service_cls: Type[RpcService], base_method: Callable[..., Any]) -> None:
+        super().__init__()
+        self._base_service_cls = base_service_cls
+        self._base_method = base_method
+        self._model = self._create_pydantic_model()
+
+    @property
+    def service_name(self) -> str:
+        return self._base_service_cls.__name__
+
+    @property
+    def method_name(self) -> str:
+        return self._base_method.__name__
+
+    def _create_pydantic_model(self) -> Type[pydantic.BaseModel]:
+        def create_field(param: inspect.Parameter) -> Tuple[Any, Any]:
+            if param.annotation is param.empty:
+                raise RpcServiceSetupException("Type hints are required on RPC methods")
+
+            default_value = ... if param.default is param.empty else param.default
+            return param.annotation, default_value
+
+        name = f"{self.service_name}__{self.method_name}__ParameterModel"
+        parameters = list(inspect.signature(self._base_method).parameters.values())
+        parameters = parameters[1:]  # exclude `self` argument
+        field_definitions = {p.name: create_field(p) for p in parameters}
+        return pydantic.create_model(name, **field_definitions)  # type: ignore
+
+    def serialize_arguments(self, raw_arguments: ArgumentDict) -> ArgumentDict:
+        model_instance = self._model(**raw_arguments)
+        return model_instance.dict()
+
+    def deserialize_arguments(self, serial_arguments: ArgumentDict) -> pydantic.BaseModel:
+        return self._model.parse_obj(serial_arguments)
+
+    def resolve_to_region(self, arguments: ArgumentDict) -> Region:
+        if self._base_service_cls.local_mode != SiloMode.REGION:
+            raise RpcServiceSetupException(f"{self.service_name} does not run on the region silo")
+
+        raise NotImplementedError()  # TODO
```

Comment:
If it would be better for us to actually change signatures and rely on some explicit signal (like always organization_id or region_name), we can accomodate by changing existing services.

Reply:
I'm exploring this in #45448.

Reply:
I'll review that next.